sympy/solvers/ode/single.py

```diff
@@ -359,6 +358,143 @@ def inverse(self):
+    r"""
+    Solves 1st order exact ordinary differential equations.
+
+    A 1st order differential equation is called exact if it is the total
+    differential of a function. That is, the differential equation
+
+    .. math:: P(x, y) \,\partial{}x + Q(x, y) \,\partial{}y = 0
+
+    is exact if there is some function `F(x, y)` such that `P(x, y) =
+    \partial{}F/\partial{}x` and `Q(x, y) = \partial{}F/\partial{}y`.  It can
+    be shown that a necessary and sufficient condition for a first order ODE
+    to be exact is that `\partial{}P/\partial{}y = \partial{}Q/\partial{}x`.
+    Then, the solution will be as given below::
+
+        >>> from sympy import Function, Eq, Integral, symbols, pprint
+        >>> x, y, t, x0, y0, C1= symbols('x,y,t,x0,y0,C1')
+        >>> P, Q, F= map(Function, ['P', 'Q', 'F'])
+        >>> pprint(Eq(Eq(F(x, y), Integral(P(t, y), (t, x0, x)) +
+        ... Integral(Q(x0, t), (t, y0, y))), C1))
+                    x                y
+                    /                /
+                   |                |
+        F(x, y) =  |  P(t, y) dt +  |  Q(x0, t) dt = C1
+                   |                |
+                  /                /
+                  x0               y0
+
+    Where the first partials of `P` and `Q` exist and are continuous in a
+    simply connected region.
+
+    A note: SymPy currently has no way to represent inert substitution on an
+    expression, so the hint ``1st_exact_Integral`` will return an integral
+    with `dy`.  This is supposed to represent the function that you are
+    solving for.
+
+    Examples
+    ========
+    >>> from sympy import Symbol,Function
+    >>> from sympy import E,dsolve,Eq,log,sin,cos
+    >>> x=Symbol('x')
+    >>> f=Function('f')
+    >>> eq=Eq(0,(sin(x)*sin(f(x))-x*E**f(x))*f(x).diff(x)-(E**f(x)+cos(x)*cos(f(x))))
+    >>> dsolve(eq,hint='1st_exact')
+    Eq(x*exp(f(x)) + sin(x)*cos(f(x)), C1)
+    >>> dsolve(cos(f(x)) - (x*sin(f(x)) - f(x)**2)*f(x).diff(x),f(x), hint='1st_exact')
+    Eq(x*cos(f(x)) + f(x)**3/3, C1)
+
+    References
+    ==========
+
+    - "Differential Equations With Applications And Historical Notes",
+       Second Edition, George.F.Simmons
+    - https://en.wikipedia.org/wiki/Exact_differential_equation
+
+    # indirect doctest
+
+    """
+    hint = "1st_exact"
+    has_integral = True
+    order = [1]
+
+    def _wilds(self, f, x, order):
+        P = Wild('P', exclude=[f(x).diff(x)])
+        Q = Wild('Q', exclude=[f(x).diff(x)])
+        return P, Q
+
+    def _equation(self, fx, x, order):
+        P, Q = self.wilds()
+        return P + Q*fx.diff(x)
+
+    def _verify(self, fx) -> bool:
+        P, Q = self.wilds()
+        x = self.ode_problem.sym
+        y = Dummy('y')
+
+        m, n = self.wilds_match()
+
+        try:
+            if simplify(m) != 0:
+                m = m.subs(fx,y)
+                n = n.subs(fx,y)
+                numerator= simplify(m.diff(y)-n.diff(x))
```

Comment:
I still think that all of this code can be simplified somehow. Are all these calls to `simplify` necessary? In general we try to avoid calling `simplify` in library code because it can be slow. It is better to call specific simplify functions like `factor`, `cancel` etc. if we know what is needed.

Reply:
Maybe we can just use something like `if m.is_zero` rather than `if simplify(m) != 0`.

Reply:
`simplify` is called only before critical comparisons for converting to exact ODE, therefore removing doesn't seem possible. `is_zero` would not work here because it doesn't simplify expression before returning. I think this is the best way to simplify at present.

Reply:
I would rather not call simplify during the matching code so it might be better just to be inconclusive in some cases.

Can you give an example of what `m` and `n` might be here so that `simplify` is needed?

Reply:
'is_zero` won't work here because it doesn't simplify the expression before return. Here, `simplify` is called only before critical conditions to convert to exact ODE. At present I think `simplify` is the best way to simplify expressions here.

Reply:
Can you give the example of what `m` and `n` might be? There is almost always a better option then using generic `simplify`.